core/network-libp2p/src/custom_proto/behaviour.rs

```diff
@@ -0,0 +1,471 @@
+			peer_id: target.clone(),
+			event: CustomProtosHandlerIn::SendCustomMessage {
+				protocol: protocol_id,
+				data: data.into(),
+			}
+		});
+	}
+
+	/// Updates the attempted connections to nodes.
+	///
+	/// Also updates `next_connect_to_nodes` with the earliest known moment when we need to
+	/// update connections again.
+	fn connect_to_nodes(&mut self, params: &mut PollParameters<NetTopology>) {
+		// Make sure we are connected or connecting to all the reserved nodes.
+		for reserved in self.reserved_peers.iter() {
+			// TODO: don't generate an event if we're already in a pending connection
+			if !self.enabled_peers.contains_key(&reserved) {
+				self.events.push(NetworkBehaviourAction::DialPeer { peer_id: reserved.clone() });
+			}
+		}
+
+		// We're done with reserved node; return early if there's nothing more to do.
+		if self.reserved_only {
+			return
+		}
+
+		// Counter of number of connections to open, decreased when we open one.
+		let mut num_to_open = {
+			let num_outgoing_connections = self.enabled_peers
+				.iter()
+				.filter(|(_, endpoint)| endpoint.is_dialer())
+				.filter(|(p, _)| !self.reserved_peers.contains(p))
+				.count();
+			self.max_outgoing_connections - num_outgoing_connections
+		};
+
+		trace!(target: "sub-libp2p", "Connect-to-nodes round; attempting to fill {:?} slots",
+			num_to_open);
+
+		let local_peer_id = params.local_peer_id().clone();
+		let (to_try, will_change) = params.topology().addrs_to_attempt();
+		for (peer_id, _) in to_try {
+			if num_to_open == 0 {
+				break
+			}
+
+			if peer_id == &local_peer_id {
+				continue
+			}
+
+			if let Some((_, ban_end)) = self.banned_peers.iter().find(|(p, _)| p == peer_id) {
+				if *ban_end > Instant::now() {
+					continue
+				}
+			}
+
+			num_to_open -= 1;
+			self.events.push(NetworkBehaviourAction::DialPeer { peer_id: peer_id.clone() });
+		}
+
+		// Next round is when we expect the topology will change.
+		self.next_connect_to_nodes.reset(will_change);
+	}
+}
+
+impl<TSubstream> NetworkBehaviour<NetTopology> for CustomProtos<TSubstream>
+where
+	TSubstream: AsyncRead + AsyncWrite,
+{
+	type ProtocolsHandler = CustomProtosHandler<TSubstream>;
+	type OutEvent = CustomProtosOut;
+
+	fn new_handler(&mut self) -> Self::ProtocolsHandler {
+		CustomProtosHandler::new(self.registered_protocols.clone())
+	}
+
+	fn inject_connected(&mut self, peer_id: PeerId, endpoint: ConnectedPoint) {
+		// When a peer connects, its handler is initially in the disabled state. We make sure that
+		// the peer is allowed, and if so we put it in the enabled state.
+
+		let is_reserved = self.reserved_peers.contains(&peer_id);
+		if self.reserved_only && !is_reserved {
+			debug!(target: "sub-libp2p", "Ignoring {:?} because we're in reserved mode", peer_id);
+			return
+		}
+
+		// Check whether peer is banned.
+		if !is_reserved {
+			if let Some((_, expire)) = self.banned_peers.iter().find(|(p, _)| p == &peer_id) {
+				if *expire < Instant::now() {
```

Comment:
Shouldn't it be `*expire > Instant::now()`. Because when `expire < Instant::now()` the ban is over?